File: project-euler.java/src/main/java/euler/solutions/_17.java
package euler.solutions;

import euler.common.BaseSolution;

import java.util.List;

import static com.google.common.collect.Lists.newArrayList;
import static java.util.Arrays.asList;

public class _17 extends BaseSolution {

    private static final String[] NUMBER_WORDS_MAP = {null,
            "one","two","three","four","five","six","seven","eight","nine","ten",
            "eleven","twelve","thirteen","fourteen","fifteen","sixteen","seventeen","eighteen","nineteen" };

    private static final String[] NUMBER_10S_WORDS_MAP = {null, null,
            "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety" };

    @Override
    public long solve() {
        int sum = 0;
        for (int number = 1; number <= 1000; number++) {
            for (String word : numberWords(number)) {
                sum += word.length();
            }
        }
        return sum;
    }

    private static List<String> numberWords(int number) {
        if (number < 20) {
            return asList(NUMBER_WORDS_MAP[number]);
        }

        if (number < 100) {
            List<String> words = newArrayList(NUMBER_10S_WORDS_MAP[number / 10]);
            if (number % 10 != 0) {
                words.add(NUMBER_WORDS_MAP[number % 10]);
            }
            return words;
        }

        if (number < 1000) {
            List<String> words = newArrayList(NUMBER_WORDS_MAP[number / 100], "hundred");
            if (number % 100 != 0) {
                words.add("and");
                words.addAll(numberWords(number % 100));
            }
            return words;
        }

        return asList("one", "thousand");
    }
}

File: project-euler.go/solutions/_17/17.go
package _17

var NUMBER_WORDS_MAP     []string
var NUMBER_10S_WORDS_MAP []string

func init() {
	NUMBER_WORDS_MAP = []string{ "",
		"one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten",
		"eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen" }

	NUMBER_10S_WORDS_MAP = []string{ "", "",
		"twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety" }
}

func Solve() int64 {
	sum := 0
	for number := 1; number <= 1000; number++ {
		for _,word := range numberWords(number) {
			sum += len(word)
		}
	}
	return int64(sum)
}

func numberWords(number int) []string {
	if number < 20 {
		return []string{ NUMBER_WORDS_MAP[number] }
	}

	if number < 100 {
		words := []string{ NUMBER_10S_WORDS_MAP[number / 10] }
		if number % 10 != 0 {
			words = append(words, NUMBER_WORDS_MAP[number % 10])
		}
		return words
	}

	if number < 1000 {
		words := []string{ NUMBER_WORDS_MAP[number / 100], "hundred" }
		if number % 100 != 0 {
			words = append(words, "and")
			words = append(words, numberWords(number % 100)...)
		}
		return words
	}

	return []string{ "one", "thousand" }
}
